packages/connection/tests/php/test-class-nonce-handler.php

```diff
@@ -0,0 +1,178 @@
+<?php // phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
+/**
+ * The nonce handler tests.
+ *
+ * @package automattic/jetpack-connection
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+use PHPUnit\Framework\TestCase;
+
+/**
+ * The nonce handler tests.
+ */
+class Test_Nonce_Handler extends TestCase {
+
+	/**
+	 * The nonce timestamp.
+	 */
+	const TIMESTAMP = '1598639691';
+
+	/**
+	 * The nonce.
+	 */
+	const NONCE = 'rAnDoM';
+
+	/**
+	 * Reset the environment after each test.
+	 */
+	public function tearDown() {
+		parent::tearDown();
+
+		Nonce_Handler::invalidate_request_nonces();
+
+		remove_action( 'shutdown', array( Nonce_Handler::class, 'clean_runtime' ) );
+	}
+
+	/**
+	 * Testing the nonce adding functionality.
+	 */
+	public function test_add() {
+		// Confirm that the nonce gets added.
+		self::assertTrue( Nonce_Handler::add( static::TIMESTAMP, static::NONCE ) );
+
+		// Confirm that the nonce is loaded from cache and still valid during the request.
+		self::assertTrue( Nonce_Handler::add( static::TIMESTAMP, static::NONCE ) );
+	}
+
+	/**
+	 * Testing the shutdown cleanup hook.
+	 */
+	public function test_shutdown_cleanup() {
+		global $wp_filter;
+
+		// Using a custom function to avoid collisions with existing callbacks.
+		$return_false = function() {
+			return false;
+		};
+
+		add_filter( 'jetpack_connection_add_nonce_cleanup', $return_false );
+		Nonce_Handler::add( static::TIMESTAMP, static::NONCE );
+		$is_callback_missing = empty( $wp_filter['shutdown']->callbacks[10][ Nonce_Handler::class . '::clean_runtime' ] );
+		remove_filter( 'jetpack_connection_add_nonce_cleanup', $return_false );
+		Nonce_Handler::invalidate_request_nonces();
+
+		Nonce_Handler::add( static::TIMESTAMP, static::NONCE );
+		$is_callback_placed = ! empty( $wp_filter['shutdown']->callbacks[10][ Nonce_Handler::class . '::clean_runtime' ] );
+
+		self::assertTrue( $is_callback_missing, 'Cleanup callback is in place, it should not exist' );
+		self::assertTrue( $is_callback_placed, 'Cleanup callback is missing, it should have been added' );
+	}
+
+	/**
+	 * Trying to add an existing nonce, and making sure it's considered invalid (returns false).
+	 */
+	public function test_add_existing() {
+		$query_filter_run = false;
+
+		$nonce_object = array(
```

Comment:
Should you cast to object here? None of the code currently cares what's actually returned from the query, but it'd be nice to be accurate.

Reply:
Done 👍